[llvm] fix nullptr dereference in BasicBlock::getIrrLoopHeaderWeight

arromanoff opened the pull request:
Currently if you try to call `getIrrLoopHeaderWeight` on a `BasicBlock` that does not have any terminators variable `TI` will be equal to nullptr and you will get an undefined behaviour from dereferencing it. This commit adds a check for this pointer and returns `std::nullopt` if no terminators found.

nikic reviewed (changes requested):
Can you please explain why this API should be expected to work with a partially constructed block? I suspect this is trying to paper over an issue elsewhere.

arrv-sc commented:
@nikic, thank you for taking a look. The main goal of this PR is to avoid UB. I don't necessarily mean that we should treat BBs with no terminators as any other blocks. But I do indeed think that causing UB in this case is not a good thing. Although IMO there's nothing inherently wrong with creating a MachineBasicBlock for incomplete BasicBlock (but I may be mistaken). In any case,  returning `std::nullopt` from function that already returns `std::optional` sounds reasonable to me

nikic commented:
> @nikic, thank you for taking a look. The main goal of this PR is to avoid UB. I don't necessarily mean that we should treat BBs with no terminators as any other blocks. But I do indeed think that causing UB in this case is not a good thing.

It would be fine to add an `assert(TI && "Block must have terminator");` to make the pre-condition explicit.

> Although IMO there's nothing inherently wrong with creating a MachineBasicBlock for incomplete BasicBlock (but I may be mistaken).

Creating a MachineBasicBlock for an incomplete basic block definitely doesn't make sense. Incomplete basic blocks exist temporarily during IR construction, they should never make it to the backend.

Commit:
[llvm] fix nullptr dereference in BasicBlock::getIrrLoopHeaderWeight
Currently if you try to call `getIrrLoopHeaderWeight` on a `BasicBlock`
that does not have any terminators variable `TI` will be equal to
nullptr and you will get an undefined behaviour from dereferencing it.
This commit adds a check for this pointer and returns `std::nullopt` if
no terminators found.

arrv-sc commented:
I think you are right that we should not guarantee that MachineBasicBlock can be constructed for an empty BasicBlock. It indeed has very little sense from compiler infrastructure point of view (will remove the test for it). Thank you. However I think it does make sense to return `std::nullopt` in the case terminator is not found.  It seems legal to me to attempt to get Loop Header Weight of the BasicBlock with no terminators (and the result should be `std::nullopt` because block is indeed incomplete and we can not analyze any loop-related stuff on WIP basic block). In case you are still strongly against handling this case let me know and I'll simply add an assert instead.

nikic commented:
> I think you are right that we should not guarantee that MachineBasicBlock can be constructed for an empty BasicBlock. It indeed has very little sense from compiler infrastructure point of view (will remove the test for it). Thank you. However I think it does make sense to return `std::nullopt` in the case terminator is not found. It seems legal to me to attempt to get Loop Header Weight of the BasicBlock with no terminators (and the result should be `std::nullopt` because block is indeed incomplete and we can not analyze any loop-related stuff on WIP basic block). In case you are still strongly against handling this case let me know and I'll simply add an assert instead.

I'd prefer the assert. I don't think querying this on an incomplete block is a legal operation. It's better to crash and burn than silently accept unexpected state.
